pyccel/codegen/printing/cucode.py

```diff
@@ -61,6 +64,83 @@ def _print_Module(self, expr):
         self.exit_scope()
         return code
 
+    def function_signature(self, expr, print_arg_names = True):
+        """
+        Get the Cuda representation of the function signature.
+
+        Extract from the function definition `expr` all the
+        information (name, input, output) needed to create the
+        function signature and return a string describing the
+        function.
+        This is not a declaration as the signature does not end
+        with a semi-colon.
+
+        Parameters
+        ----------
+        expr : FunctionDef
+            The function definition for which a signature is needed.
+
+        print_arg_names : bool, default : True
+            Indicates whether argument names should be printed.
+
+        Returns
+        -------
+        str
+            Signature of the function.
+        """
+        arg_vars = [a.var for a in expr.arguments]
+        result_vars = [r.var for r in expr.results if not r.is_argument]
+        n_results = len(result_vars)
+
+        if n_results == 1:
+            ret_type = self.get_declare_type(result_vars[0])
+        elif n_results > 1:
+            ret_type = self.find_in_dtype_registry(PythonNativeInt())
+            arg_vars.extend(result_vars)
+            self._additional_args.append(result_vars)
+        else:
+            ret_type = self.find_in_dtype_registry(VoidType())
+
+        name = expr.name
+        if not arg_vars:
+            arg_code = 'void'
+        else:
+            def get_arg_declaration(var):
+                """ Get the code which declares the argument variable.
+                """
+                code = "const " * var.is_const
+                code += self.get_declare_type(var)
+                if print_arg_names:
+                    code += ' ' + var.name
+                return code
+
+            arg_code_list = [self.function_signature(var, False) if isinstance(var, FunctionAddress)
+                                else get_arg_declaration(var) for var in arg_vars]
+            arg_code = ', '.join(arg_code_list)
+
+        if self._additional_args :
+            self._additional_args.pop()
+
+        static = 'static ' if expr.is_static else ''
+        cuda_decorater = ""
+        if('kernel' in expr.decorators):
+            cuda_decorater = "__global__"
+        return f'{static} {cuda_decorater} {ret_type} {name}({arg_code})'
+
+    def _print_KernelCall(self, expr):
+        func = expr.funcdef
+        args = []
+        for a in expr.args:
+            arg_val = a.value or Nil()
+            args.append(arg_val)
+
+        args += self._temporary_args
+        self._temporary_args = []
+        args = ', '.join([f'{self._print(a)}' for a in args])
```

Comment:
You can pass a generator to `join` to avoid allocating memory for the list.
You also seem to be using a f-string to convert a string to a string here.
```suggestion
        args = ', '.join(self._print(a) for a in args)
```

Reply:
good optimization , [b000bd9](https://github.com/pyccel/pyccel-cuda/pull/42/commits/b000bd9ea6b6a330ce2298add96f90625548a81a)